content/blog/deep-dive-merged-fields.md

```diff
@@ -0,0 +1,219 @@
++++
+title = "GraphQL Deep Dive - Part 1: merged fields"
+author = "Andreas Marek"
+tags = []
+categories = []
+date = 2019-01-22T00:00:00+10:00
++++
+
+# GraphQL Deep Dive series 
+
+Welcome to the new series "GraphQL deep dive" where we will explore advanced or unknown GraphQL topics. The plan is to discuss things mostly in a language and implementation neutral way, even if it is hosted on graphql-java.com. 
+
+# Merged Fields
+
+First thing we are looking at is "merged fields".
+
+GraphQL allows for a field to be declared multiple times in a query as long as it can be merged.
+
+Valid GraphQL queries are:
+
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo
+          foo
+        }
+{{< / highlight >}}
+
+<p/>
+
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo(id: "123")
+          foo(id: "123")
+          foo(id: "123")
+        }
+{{< / highlight >}}
+<p/>
+
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo(id: "123") {
+            id
+          }
+          foo(id: "123") {
+            name
+          }
+          foo(id: "123") {
+            id 
+            name
+          }
+        }
+{{< / highlight >}}
+
+Each of these queries will result in a result with just one "foo" key, not two or three. 
+
+Invalid Queries are:
+
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo
+          foo(id: "123")
+        }
+{{< / highlight >}}
+<p/>
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo(id: "123")
+          foo(id: "456", id2: "123")
+        }
+{{< / highlight >}}
+<p/>
+{{< highlight graphql "linenos=table" >}}
+        {
+          foo(id: "123")
+          foo: foo2
+        }
+{{< / highlight >}}
+
+The reason whey they are not valid, is because the fields are different: in the first two examples the arguments differ and the third query actually has two different fields under the same key.
+
+# Motivation
+
+The examples so far don't seem really useful, but it all makes sense when you add fragments:
+
+{{< highlight graphql "linenos=table" >}}
+        {
+          ...myFragment1
+          ...myFragment2
+        }
+
+        fragment myFragment1 on Query {
+          foo(id: "123") {
+            name
+          }
+        }
+        fragment myFragment2 on Query {
+          foo(id: "123") {
+            url
+          }
+        }
+
+{{< / highlight >}}
+
+<p/>
+Fragments are designed to be written by different parties (for example different components in a UI) which should not know anything about each other. Requiring that every field can only be declared once would make this objective unfeasible.  
+
+But by allowing it as long as the fields are the same allows fragments to be authored in an independent from each other.  
+
```

Comment:
But by allowing fields merging, as long as the fields are the same, allows fragments to be authored in an independent from each other.